org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/LSJavaSemanticTokensProvider.java

```diff
@@ -0,0 +1,162 @@
+import org.eclipse.lsp4j.ServerCapabilities;
+
+public class LSJavaSemanticTokensProvider implements ISemanticTokensProvider {
+	
+	@Override
+	public Collection<ISemanticTokensProvider.SemanticToken> computeSemanticTokens(CompilationUnit ast) {
+		IPreferenceStore prefStore = LanguageServerPlugin.getDefault().getPreferenceStore();
+		LanguageServerJdtPlugin plugin = LanguageServerJdtPlugin.getDefault();
+		if (plugin == null) {
+			throw new IllegalStateException("Plugin hasn't been started!");
+		}
+		IPreferenceStore jstPrefStore = plugin.getPreferenceStore();
+		
+		if (prefStore.getBoolean(SemanticHighlightReconcilerStrategy.SEMANTIC_HIGHLIGHT_RECONCILER_DISABLED)
+				|| !jstPrefStore.getBoolean(LspJdtConstants.PREF_SEMANTIC_TOKENS_SWITCH)) {
+			return Collections.emptyList();
+		}
+		
+		IResource resource = ast.getTypeRoot().getResource();
+		if (resource == null) {
+			return Collections.emptyList();
+		}
+		
+		final IDocument theDocument = LSPEclipseUtils.getDocument(resource);
+		if (theDocument == null) {
+			return Collections.emptyList();
+		}
+		
+		SemanticTokensParams params = getSemanticTokensParams(theDocument);
+		if (params == null) {
+			return Collections.emptyList();
+		}
+		
+		LanguageServerDocumentExecutor executor = LanguageServers.forDocument(theDocument)
+				.withFilter(this::hasSemanticTokensFull);
+			
+		try {
+			return executor
+				.computeFirst((w, ls) -> {
+					return ls.getTextDocumentService().semanticTokensFull(params)
+							.thenApply(semanticTokens -> {
+								if (semanticTokens == null) {
+									return Collections.<SemanticToken>emptyList();
+								}
+								SemanticTokensLegend legend = getSemanticTokensLegend(w);
+								if (legend == null) {
+									return Collections.<SemanticToken>emptyList();
+								}
+								return new JavaSemanticTokensProcessor(this::mapToTokenType, p -> {
+									try {
+										return LSPEclipseUtils.toOffset(p, theDocument);
+									} catch (BadLocationException e) {
+										throw new RuntimeException(e);
+									}
+								}).getSemanticTokens(semanticTokens.getData(), legend);
+							});
+				})
+				.thenApply(o -> o.orElse(Collections.emptyList())).get(300, TimeUnit.MILLISECONDS);
+		} catch (InterruptedException | ExecutionException | TimeoutException e) {
+			LanguageServerPlugin.logError("Failed to fetch semantic tokens for '%s' from Language Servers".formatted(resource.getLocation()), e);
+			return Collections.emptyList();
+		}		
+	}
+	
+	private @Nullable SemanticTokensParams getSemanticTokensParams(IDocument document) {
+		URI uri = LSPEclipseUtils.toUri(document);
+		if (uri != null) {
+			final var semanticTokensParams = new SemanticTokensParams();
+			semanticTokensParams.setTextDocument(LSPEclipseUtils.toTextDocumentIdentifier(uri));
+			return semanticTokensParams;
+		}
+		return null;
+	}
+
+	private @Nullable SemanticTokensLegend getSemanticTokensLegend(final LanguageServerWrapper wrapper) {
```

Comment:
This method (and others) is an exact copy of org.eclipse.lsp4e.operations.semanticTokens.SemanticHighlightReconcilerStrategy.getSemanticTokensLegend(LanguageServerWrapper).

Others are very similar.  Do you think we could refactor the code to avoid the redundancy?